### PR TITLE
[ML] add new inference_config field to trained model config

### PR DESCRIPTION
**Pull request by benwtrent**
A new field called `inference_config` is now added to the trained model config object. This new field allows for default inference settings from analytics or some external model builder. 

The inference processor can still override whatever is set as the default in the trained model config.

Docs preview:
* http://elasticsearch_54421.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/put-inference.html
* http://elasticsearch_54421.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/inference-processor.html
* http://elasticsearch_54421.docs-preview.app.elstc.co/guide/en/elasticsearch/client/java-rest/master/java-rest-high-x-pack-ml-put-trained-model.html

### CODE REVIEW

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Comment by benwtrent** [comment]
@elasticmachine update branch

File: docs/java-rest/high-level/ml/put-trained-model.asciidoc [hunk 1]
```diff
@@ -39,6 +39,8 @@ include-tagged::{doc-tests-file}[{api}-config]
 <5> Optionally, a human-readable description
 <6> Optionally, an object map contain metadata about the model
 <7> Optionally, an array of tags to organize the model
+<8> The default inference config to use with the model. Must match the underlying
```
Comment (lcawl) [inline docs/java-rest/high-level/ml/put-trained-model.asciidoc]: I think the new object needs to be added in the definitions section too (e.g. https://www.elastic.co/guide/en/elasticsearch/reference/master/put-inference.html#ml-put-inference-trained-model)

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/AnalyticsResultProcessor.java [hunk 2]
```diff
@@ -231,9 +235,37 @@ private TrainedModelConfig createTrainedModelConfig(TrainedModelDefinition.Build
...
+                    ClassificationConfig.DEFAULT_TOP_CLASSES_RESULTS_FIELD,
+                    classification.getBoostedTreeParams().getNumTopFeatureImportanceValues() == null ?
+                        0 :
+                        classification.getBoostedTreeParams().getNumTopFeatureImportanceValues());
+            case REGRESSION:
+                assert analytics.getAnalysis() instanceof Regression;
+                Regression regression = ((Regression)analytics.getAnalysis());
+                return new RegressionConfig(RegressionConfig.DEFAULT_RESULTS_FIELD,
+                    regression.getBoostedTreeParams().getNumTopFeatureImportanceValues() == null ?
+                        0 :
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/AnalyticsResultProcessor.java:258]: Let `RegressionConfig` pick the default value (it does this in the constructor anyway)

```
RegressionConfig(RegressionConfig.DEFAULT_RESULTS_FIELD,
                    regression.getBoostedTreeParams().getNumTopFeatureImportanceValues())
```

Comment (davidkyle) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/AnalyticsResultProcessor.java:258]: Same with `ClassificationConfig`

**Review by davidkyle — APPROVED** [review]
Great change! LGTM

One request could we now make the inference config for the inference processor optional? i.e. remove the need for 
```
"inference_config": {
    "regression": {}
}
```

File: client/rest-high-level/src/test/java/org/elasticsearch/client/ml/inference/trainedmodel/ClassificationConfigTests.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.test.AbstractXContentTestCase;
+
+import java.io.IOException;
+
+public class ClassificationConfigTests extends AbstractXContentTestCase<ClassificationConfig> {
+
+    public static ClassificationConfig randomClassificationConfig() {
+        return new ClassificationConfig(randomBoolean() ? null : randomIntBetween(-1, 10),
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/ml/inference/trainedmodel/ClassificationConfigTests.java:30]: Yes, indicates "_all"

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ClassificationConfigUpdate.java [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        return apply(ClassificationConfig.EMPTY_PARAMS);
+    }
+
+    @Override
+    public boolean isSupported(InferenceConfig inferenceConfig) {
+        return inferenceConfig instanceof ClassificationConfig;
+    }
+
+    boolean isNoop(ClassificationConfig originalConfig) {
+        return (resultsField == null || originalConfig.getResultsField().equals(resultsField))
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ClassificationConfigUpdate.java:198]: roger

**Comment by benwtrent** [comment]
> One request could we now make the inference config for the inference processor optional?

Part of me is still reticent to do that. It is right now, the only indication of the "inference type". I suppose we can remove it in the future for sure. I don't think we should do it in this PR though.

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Review by lcawl — APPROVED** [review]
Documentation LGTM

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Pull request merged**